### PR TITLE
Profiler overhead measurement

### PR DESCRIPTION
**Pull request by white-pony**
I do not think this PR is fully ready (I apologize), but I wanted to ask about your opinion.

I've been trying to improve precision of overhead adjustment (as it sometimes leads to negative execution time values) and initially, for the overhead measurement,  it generates the following code (eight calls to profiling_timer()):
```c++
for (j, 0, 1000000) {
  let begin_halide_profiler_ticks_dst_$ignore$_$ignore$_null_null = profiling_timer()
  $ignore_buf$[0] = uint32(0)
  ProfilerBuffer[13] = (ProfilerBuffer[13] + (profiling_timer() - begin_halide_profiler_ticks_dst_$ignore$_$ignore$_null_null))
  let begin_halide_profiler_ticks_dst_$ignore$_$ignore$_null_null = profiling_timer()
  $ignore_buf$[0] = uint32(0)
  ProfilerBuffer[13] = (ProfilerBuffer[13] + (profiling_timer() - begin_halide_profiler_ticks_dst_$ignore$_$ignore$_null_null))
  let begin_halide_profiler_ticks_dst_$ignore$_$ignore$_null_null = profiling_timer()
  $ignore_buf$[0] = uint32(0)
  ProfilerBuffer[13] = (ProfilerBuffer[13] + (profiling_timer() - begin_halide_profiler_ticks_dst_$ignore$_$ignore$_null_null))
  let begin_halide_profiler_ticks_dst_$ignore$_$ignore$_null_null = profiling_timer()
  $ignore_buf$[0] = uint32(0)
  ProfilerBuffer[13] = (ProfilerBuffer[13] + (profiling_timer() - begin_halide_profiler_ticks_dst_$ignore$_$ignore$_null_null))
}
```
which gets simplified into (five calls to profiling_timer()):
```c++
for (j, 0, 1000000) {
  let begin_halide_profiler_ticks_dst_$ignore$_$ignore$_null_null = profiling_timer()
  $ignore_buf$[0] = uint32(0)
  ProfilerBuffer[13] = (ProfilerBuffer[13] + (profiling_timer() - begin_halide_profiler_ticks_dst_$ignore$_$ignore$_null_null))
  $ignore_buf$[0] = uint32(0)
  ProfilerBuffer[13] = (ProfilerBuffer[13] + (profiling_timer() - begin_halide_profiler_ticks_dst_$ignore$_$ignore$_null_null))
  $ignore_buf$[0] = uint32(0)
  ProfilerBuffer[13] = (ProfilerBuffer[13] + (profiling_timer() - begin_halide_profiler_ticks_dst_$ignore$_$ignore$_null_null))
  $ignore_buf$[0] = uint32(0)
  ProfilerBuffer[13] = (ProfilerBuffer[13] + (profiling_timer() - begin_halide_profiler_ticks_dst_$ignore$_$ignore$_null_null))
}
```
As suggested in the #655, dummy argument passed to profiling_timer() prevents simplification of this code, so at the end, it turns into (again eight calls to profiling_timer()):
```c++
  for (j, 0, 1000000) {
    let begin_halide_profiler_ticks_dst_$ignore$_$ignore$_null_null = profiling_timer((dummy + 9))
    $ignore_buf$[0] = uint32(0)
    ProfilerBuffer[13] = (ProfilerBuffer[13] + (profiling_timer((dummy + 9)) - begin_halide_profiler_ticks_dst_$ignore$_$ignore$_null_null))
    let begin_halide_profiler_ticks_dst_$ignore$_$ignore$_null_null = profiling_timer((dummy + 8))
    $ignore_buf$[0] = uint32(0)
    ProfilerBuffer[13] = (ProfilerBuffer[13] + (profiling_timer((dummy + 8)) - begin_halide_profiler_ticks_dst_$ignore$_$ignore$_null_null))
    let begin_halide_profiler_ticks_dst_$ignore$_$ignore$_null_null = profiling_timer((dummy + 7))
    $ignore_buf$[0] = uint32(0)
    ProfilerBuffer[13] = (ProfilerBuffer[13] + (profiling_timer((dummy + 7)) - begin_halide_profiler_ticks_dst_$ignore$_$ignore$_null_null))
    let begin_halide_profiler_ticks_dst_$ignore$_$ignore$_null_null = profiling_timer((dummy + 6))
    $ignore_buf$[0] = uint32(0)
    ProfilerBuffer[13] = (ProfilerBuffer[13] + (profiling_timer((dummy + 6)) - begin_halide_profiler_ticks_dst_$ignore$_$ignore$_null_null))
  }
```
Now, in order to calculate overhead (let call it P), total execution time needs to be divided by number of iterations (1000000 in this case) multiplied by number of measurements (4 in this case). However, this will give correct estimation only for inner loops/ops (i.e. for children), because it includes two calls to profiling_timer (here I am making assumption that profiling_timer() dominates everything else). However, in order, to get adjustment for loop/ops itself, P should be divided by two, as it includes only one call to profiling_timer(). So, my point is that children values should be adjusted differently. Does it make sense?

I am not sure about this one, but also, it seems that putting profiling around loop which is going to be vectorized (i.e. one line of SIMD) does not make much sense, because usually execution of it takes a really small amount of time and it's significantly smaller than one for profiling code.

### CODE REVIEW

File: src/Profiling.cpp [hunk 1]
```diff
@@ -159,7 +165,7 @@ class InjectProfiling : public IRMutator {
     }
 
     Stmt add_ticks(const string& op_type, const string& op_name, Stmt s) {
-        Expr ticks = Call::make(UInt(64), Internal::Call::profiling_timer, vector<Expr>(), Call::Intrinsic);
+        Expr ticks = Call::make(UInt(64), Internal::Call::profiling_timer, {dummy + dummy_counter++}, Call::Intrinsic);
```
Comment (steven-johnson) [inline src/Profiling.cpp]: We still have some builds that don't require C++11.

**Comment by steven-johnson** [comment]
I'm not sure I understand your point -- the example above is (I presume) trying just to measure the average overhead of a single call to profiling_timer()?

**Comment by white-pony** [comment]
Sorry for messy explanation. My point is that two different cases are happening, so they needed to be adjusted slightly differently.

For example, if we want to adjust overhead for the *outer_total_time* from the following example:
```c++
let outer_start_time = profiling_timer((dummy + 1))
let outer_start_count = uint64(0)
for (dst.s0.x.xo, dst.s0.x.xo.loop_min, dst.s0.x.xo.loop_extent) {
    let inner_start_time = profiling_timer((dummy + 0))
    let inner_start_count = uint64(0)
    for (dst.s0.x.x, dst.s0.x.x.loop_min, dst.s0.x.x.loop_extent) {
        let dst.s0.x = (dst.s0.x.x.base + dst.s0.x.x)
        dst((dst.s0.x.x.base + dst.s0.x.x)) = src((dst.s0.x.x.base + dst.s0.x.x))
    }
    inner_total_count = (inner_total_count + (uint64(1) - inner_start_count))
    inner_total_time = (inner_total_time + (profiling_timer((dummy + 0)) - inner_start_time))
}
outer_total_count = (outer_total_count + (uint64(1) - outer_start_count))
outer_total_time = (outer_total_time + (profiling_timer((dummy + 1)) - outer_start_time))
```
and overhead is calculated as an execution time of the following block (let's call it P):
```c++
let start_time = profiling_timer((dummy + 8))
ProfilerBuffer[11] = (ProfilerBuffer[11] + (profiling_timer((dummy + 8)) - start_time))
```
 we will sum all counts for outer and inner level, multiply it by P and subtract result from the total:
*outer_total_time_adj = outer_total_time - P * outer_total_count - P * inner_total_count*. 

My point is that overhead for outer_* and inner_* is different. It's P for the inner loop and P/2 for the outer, because outer_time includes only half of profiling block's time (approximately, assuming that profiling_timer() dominates). So overall, adjustment turns into:
*outer_total_time_adj = outer_total_time - P / 2 * outer_total_count - P * inner_total_count*

File: src/Profiling.cpp (hunk 1 above)
Comment (white-pony) [inline src/Profiling.cpp]: Fixed.

**Comment by steven-johnson** [comment]
Ah, right, that makes sense, and seems correct. 

You said this PR isn't fullyr ready; please let us know when you feel it is ready for a proper review.

**Comment by white-pony** [comment]
Ok, great. just wanted to make sure that I'm getting it right.

Thanks.

**Comment by white-pony** [comment]
It should be ready now. Please, take a look.

**Comment by white-pony** [comment]
I've added comment and split if condition into two parts (it appears to me to be more readable).

**Comment by steven-johnson** [comment]
(I just noticed that there don't appear to be any tests for the profiler in the standard test suite. I'll open an issue that we probably should add some...)

**Comment by steven-johnson** [comment]
Merged

**Pull request merged**